kratos/python_scripts/sympy_fe_utilities.py

```diff
@@ -1,466 +1,522 @@
+    - expression -- The expression to geneate code from
+    - language -- The language of output
+    - indentation_level -- The number of tabulations considered
+    - max_index -- The maximum index
+    - replace_indices -- Set to `True` to replace matrix[i,j] with matrix_i_j (And similarly for vectors)
     """
-    initial_spaces = str("")
-    for i in range(0,initial_tabs):
-        initial_spaces += str("    ")
+    outstring = _CodeGen(language, expression) + _Suffix(language)
 
-    outstring = str("")
-    if(mode == "python"):
-        outstring += initial_spaces+str(var)+str("\n")
-    elif(mode=="c"):
-        outstring += initial_spaces+ccode(var)+str(";\n")
+    if replace_indices:
+        outstring = _ReplaceIndices(language, outstring)
+
+    return outstring
+
+def OutputSymbolicVariableDeclaration(expression, name, language, indentation_level=0, replace_indices=True):
+    """
+    This function generates code to declare and assign an expression, such as:
+
+    ```C++
+        const double variable = expression;
+
+    ```
+
+    Keyword arguments:
+    - expression -- The variable to define symbolic
+    - language -- The language of output
+    - name -- The name of the variables
+    - indentation_level -- The number of tabulations considered
+    - max_index -- DEPRECATED The maximum index
+    - replace_indices -- Set to `True` to replace matrix[i,j] with matrix_i_j (And similarly for vectors)
+    """
+    prefix = _Indentation(indentation_level)
+    value = _CodeGen(language, expression)
+    expr = _VariableDeclaration(language, name, value)
+    suffix = _Suffix(language)
+
+    outstring = prefix + expr + suffix
 
     if replace_indices:
-        #matrix entries (two indices)
-        for i in range(0,max_index):
-            for j in range(0,max_index):
-                if(mode == "python"):
-                    replacement_string = str("[") + str(i) + str(",") + str(j) + str("]")
-                elif(mode=="c"):
-                    replacement_string = str("(") + str(i) + str(",") + str(j) + str(")")
-                to_be_replaced = str("_") + str(i) + str("_") + str(j)
-                newstring = outstring.replace(to_be_replaced, replacement_string)
-                outstring = newstring
-
-        #vector entries (one index)
-        for i in range(0,max_index):
-            replacement_string = str("[") + str(i) + str("]")
-            to_be_replaced = str("_") + str(i)
-            newstring = outstring.replace(to_be_replaced, replacement_string)
-            outstring = newstring
+        _ReplaceIndices(language, outstring)
 
     return outstring
 
-def OutputMatrix_CollectingFactors(A, name, mode, initial_tabs=3, max_index=30, optimizations='basic', replace_indices=True, assignment_op="="):
-    """ This method collects the constants of the replacement for matrices
+def _OutputX_CollectionFactors(A, name, language, indentation_level, optimizations, replace_indices, assignment_op, output_func):
```

Comment:
```suggestion
def _AuxiliaryOutputCollectionFactors(A, name, language, indentation_level, optimizations, replace_indices, assignment_op, output_func):
```
Very minor naming suggestion.